src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+}
+
+TEST_P(DepthStencilStateTest, StencilInvert) {
+    CheckStencilOperation(nxt::StencilOperation::Invert, 0xf0, 3, 0x0f);
+}
+
+TEST_P(DepthStencilStateTest, StencilIncrementClamp) {
+    CheckStencilOperation(nxt::StencilOperation::IncrementClamp, 1, 3, 2);
+    CheckStencilOperation(nxt::StencilOperation::IncrementClamp, 0xff, 3, 0xff);
+}
+
+TEST_P(DepthStencilStateTest, StencilIncrementWrap) {
+    CheckStencilOperation(nxt::StencilOperation::IncrementWrap, 1, 3, 2);
+    CheckStencilOperation(nxt::StencilOperation::IncrementWrap, 0xff, 3, 0);
+}
+
+TEST_P(DepthStencilStateTest, StencilDecrementClamp) {
+    CheckStencilOperation(nxt::StencilOperation::DecrementClamp, 1, 3, 0);
+    CheckStencilOperation(nxt::StencilOperation::DecrementClamp, 0, 3, 0);
+}
+
+TEST_P(DepthStencilStateTest, StencilDecrementWrap) {
+    CheckStencilOperation(nxt::StencilOperation::DecrementWrap, 1, 3, 0);
+    CheckStencilOperation(nxt::StencilOperation::DecrementWrap, 0, 3, 0xff);
+}
+
+TEST_P(DepthStencilStateTest, StencilReadMask) {
+    nxt::DepthStencilState baseState = device.CreateDepthStencilStateBuilder()
+        .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Replace)
+        .GetResult();
+
+    nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+        .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Equal, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep)
+        .SetStencilMask(0x2, 0xff)
+        .GetResult();
+
+    RGBA8 baseColor = RGBA8(255, 255, 255, 255);
+    RGBA8 red = RGBA8(255, 0, 0, 255);
+    RGBA8 green = RGBA8(0, 255, 0, 255);
+
+    TestSpec base = { baseState, baseColor, 0.5f, 3u };
+    DoTest({ base, { state, red, 0.f, 1u } }, baseColor);
+    DoTest({ base, { state, green, 0.f, 2u } }, green);
+}
+
+TEST_P(DepthStencilStateTest, StencilWriteMask) {
+    nxt::DepthStencilState baseState = device.CreateDepthStencilStateBuilder()
+        .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Always, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Replace)
+        .SetStencilMask(0xff, 0x1)
+        .GetResult();
+
+    nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+        .SetStencilFunction(nxt::Face::Both, nxt::CompareFunction::Equal, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep, nxt::StencilOperation::Keep)
+        .GetResult();
+
+    RGBA8 baseColor = RGBA8(255, 255, 255, 255);
+    RGBA8 green = RGBA8(0, 255, 0, 255);
+
+    TestSpec base = { baseState, baseColor, 0.5f, 3u };
+    DoTest({ base, { state, green, 0.f, 2u } }, baseColor);
+    DoTest({ base, { state, green, 0.f, 1u } }, green);
+}
+
+// Draw a base triangle and then a second triangle such that the stencil test will fail. Replace the stencil value on stencil failure
```

Comment:
nit: This comment could be inside the test, with a comment describing what this tested above the TEST_P. Same below.